examples/with-apollo/lib/withData.js

```diff
@@ -0,0 +1,44 @@
+import { ApolloProvider, getDataFromTree } from 'react-apollo'
+import React from 'react'
+import 'isomorphic-fetch'
+import { IS_SERVER } from './exenv'
+import { initClient } from './initClient'
+import { initStore } from './initStore'
+
+export default (Component) => (
+  class extends React.Component {
+    static async getInitialProps (ctx) {
+      const headers = ctx.req ? ctx.req.headers : {}
+      const client = initClient(headers)
+      const store = initStore(client, client.initialState)
+
+      if (IS_SERVER) {
+        const app = (
+          <ApolloProvider client={client} store={store}>
+            <Component url={{ query: ctx.query, pathname: ctx.pathname }} />
+          </ApolloProvider>
+        )
+        await getDataFromTree(app)
+      }
+
+      return {
+        initialState: store.getState(),
```

Comment:
Apollo only permits rehydrating `state.apollo.data`. See https://github.com/apollostack/apollo-client/issues/759

Otherwise all kind of weird bugs can happen. I found out the hard way:
When using a `ssr: false` query `state.apollo.queries` mismatched on the client and gave me a bogus warning about a missing query variable. Apollo client unfortunately uses a bad query id https://github.com/apollostack/apollo-client/blob/41a3919/src/core/QueryManager.ts#L543 which causes things to mismatch if the order of queries isn't the same on the sever and client.

I recommend changing this to the following:
```
const state = store.getState()
const initialProps = {
  initialState: {
    ...state,
    apollo: {
      data: state.apollo.data
    }
  },
  headers
}
return initialProps
```